userland/libtock/aes.c

```diff
@@ -1,37 +1,165 @@
 #include "aes.h"
 
-int aes128_init(subscribe_cb callback, void *ud) {
+// used for creating synchronous versions of functions
+//
+// fired -  set when the callback has been called
+// error - error received from the kernel less than zero indicates an error
+typedef struct {
+  bool fired;
+  int error;
+} aes_data_t;
+
+
+// Internal callback for creating synchronous functions
+//
+// callback_type - number indicating which type of callback occurred
+// currently 1(encryption) and 2(decryption)
+// callback_args - user data passed into the set_callback function
+//
+static void aes_cb(int callback_type,
+    __attribute__ ((unused)) int unused1,
+    __attribute__ ((unused)) int unused2,
+    void *callback_args) {
+
+  aes_data_t *result = (aes_data_t*)callback_args;
+  result->fired = true;
+  result->error = callback_type;
+}
+
+
+// ***** System Call Interface *****
+
+// Internal callback for encryption and decryption
+int aes128_set_callback(subscribe_cb callback, void *ud) {
   return subscribe(AES_DRIVER, 0, callback, ud);
 }
 
-int aes128_configure_key(const unsigned char* key, unsigned char len) {
-  int err = allow(AES_DRIVER, AES_KEY, (void*)key, len);
-  if (err < 0)  {
-    return err;
-  }
-  return command(AES_DRIVER, AES_KEY, 0);
+// Internal function to configure a payload to encrypt or decrypt
+int aes128_set_data(const unsigned char *data, unsigned char len) {
+  return allow(AES_DRIVER, AES_DATA, (void*)data, len);
+}
+
+// Internal function to configure a initial counter to be used on counter-mode
+int aes128_set_ctr(const unsigned char* ctr, unsigned char len) {
+  return allow(AES_DRIVER, AES_CTR, (void*)ctr, len);
 }
 
-int aes128_encrypt_ctr(unsigned const char* buf, unsigned char buf_len, unsigned const char* ctr, unsigned char ctr_len) {
-  int err = allow(AES_DRIVER, AES_DATA, (void*)buf, buf_len);
-  if (err < 0)  {
-    return err;
-  }
-  err = allow(AES_DRIVER, AES_CTR, (void*)ctr, ctr_len);
-  if (err < 0) {
-    return err;
-  }
+// Internal function to trigger encryption operation. Note that this don't
+// work by itself aes128_set_data() and aes128_set_ctr() must be called first
+// FIXME: better name?
+int aes128_encrypt(void) {
```

Comment:
I'm happy with them being named whatever you think is correct. I agree with `aes128_encrypt` being a little confusing. Maybe `aes128_start_encryption`?